### ISSUE TITLE
BiocManager does not work for the latest devel version 4.2

### ISSUE THREAD

**Issue by LiNk-NY** [Bioconductor/BiocManager#99]
As of 4/20, the R-devel version changed to `4.2.0`. This causes an error with BiocManager: 

```r
> BiocManager::install(version="3.13", update=TRUE, ask=FALSE)
Error: Bioconductor version '3.13' requires R version '4.1'; R version is too new; see
  https://bioconductor.org/install
```

A potential fix is here dfd67b2fd9ccc6578bc48ea8f1bd0ce87e59e87c which reduces this error to a (persistent) message
since `BiocManager` is supposed to work with the latest R-devel version anyway.

**Comment by hpages** [comment]
Isn't it actually an error to try to install BioC 3.13 on R 4.2?

**Comment by LiNk-NY** [comment]
It depends on what we consider R-devel. 
If we are supporting R-devel then no it shouldn't be an error.
If we are restricting `R-devel` to mean `4.1` then yes. 

We also need BiocManager to pass CRAN checks where R-devel is currently 4.2:
https://cran.r-project.org/web/checks/check_results_BiocManager.html

**Comment by hpages** [comment]
My question didn't mention R-devel, only R 4.2.

Yes, passing CRAN checks is important, but that's a different story.

**Comment by HenrikBengtsson** [comment]
FYI, today a CRAN maintainer reported that they're suffering from this problem for their package:

'[R-pkg-devel] Build fails on R-devel due to Bioconductor version mismatch', 2021-04-26 (https://stat.ethz.ch/pipermail/r-package-devel/2021q2/006898.html)

**Comment by HenrikBengtsson** [comment]
Another FYI, it also breaks R-devel checks on R-hub, regardless of the package being tested actually depend on **BiocManager** or not, e.g. https://builder.r-hub.io/status/listenv_0.8.0-9000.tar.gz-bc8e8eb24cc34c98bc93ebd19981c71e.

PS. One can probably argue that R-hub should not install **BiocManager**, if not really needed. OTH, it's probably quite tedious for them to implement such a conditional behavior.  Given that **BiocManager** is such an essential and central package, I think it should never ever break like this.  Too many side effects and too many people wasting time trying to figure out what _they_ are doing wrong.

**Comment by hpages** [comment]
Note that trying to install a version of Bioconductor that is incompatible with the running version of R has always been an error and that's probably a good thing e.g.:
```
> paste0(R.version$major, ".", R.version$minor)
[1] "4.0.3"
> library(BiocManager)
> install(version="3.13")
Error: Bioconductor version '3.13' requires R version '4.1'; use
  `BiocManager::install(version = '3.12')` with R version 4.0; see
  https://bioconductor.org/install
```
or
```
> paste0(R.version$major, ".", R.version$minor)
[1] "4.1.0"
> library(BiocManager)
> install(version="3.12")
Error: Bioconductor version '3.12' requires R version '4.0'; use
  `BiocManager::install(version = '3.13')` with R version 4.1; see
  https://bioconductor.org/install
```
I'm not sure why trying to do `BiocManager::install(version="3.13")` on R 4.2.0 doesn't produce an error message similar to the above. Maybe it should? It would probably help avoid the current confusion.

**Comment by vjcitn** [comment]
```
> BiocManager::install(version="3.13")
Error: Bioconductor version '3.13' requires R version '4.1'; R version is too new; see
  https://bioconductor.org/install
> sessionInfo()
R Under development (unstable) (2021-04-25 r80223)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Big Sur 11.2.3

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] BiocManager_1.30.12.3 devtools_2.4.0        usethis_2.0.1        

loaded via a namespace (and not attached):
 [1] rstudioapi_0.13   magrittr_2.0.1    pkgload_1.2.1     R6_2.5.0         
 [5] rlang_0.4.10      fastmap_1.1.0     tools_4.2.0       pkgbuild_1.2.0   
 [9] sessioninfo_1.1.1 cli_2.5.0         withr_2.4.2       ellipsis_0.3.1   
[13] remotes_2.3.0     rprojroot_2.0.2   lifecycle_1.0.0   crayon_1.4.1     
[17] processx_3.5.1    purrr_0.3.4       callr_3.7.0       fs_1.5.0         
[21] ps_1.6.0          curl_4.3          testthat_3.0.2    memoise_2.0.0    
[25] glue_1.4.2        cachem_1.0.4      compiler_4.2.0    desc_1.3.0       
[29] prettyunits_1.1.1```

**Comment by vjcitn** [comment]
Also of interest:
```
> tail(BiocManager:::.version_map())
   Bioc   R  BiocStatus RSPM MRAN
29  3.9 3.6 out-of-date <NA> <NA>
30 3.10 3.6 out-of-date <NA> <NA>
31 3.11 4.0 out-of-date <NA> <NA>
32 3.12 4.0     release <NA> <NA>
33 3.13 4.1       devel <NA> <NA>
34 3.13 4.2      future <NA> <NA>
```

**Comment by LiNk-NY** [comment]
Hi Vince @vjcitn and Hervé @hpages, 

I've been working on a `future` branch for `BiocManager` which would allow compatibility for cases like this. It converts the error to a persistent message when the R version is in the `future` row (see map above).

branch:
https://github.com/Bioconductor/BiocManager/tree/future

diff:
https://github.com/Bioconductor/BiocManager/compare/master..future

**Comment by vjcitn** [comment]
It's an interesting approach.  Should be discussed by entire team, maybe at TAB.

**Comment by hpages** [comment]
@vjcitn 

> It does:
> ...

Yes, it produces an error but note that the error message is not as clear as when you try to do `install(version="3.13")` on R 4.0 or `install(version="3.12")` on R 4.1. In particular it doesn't mention the version of R that's being used: it only says "too new".

@LiNk-NY It seems that some people recently installed R-devel thinking they were getting 4.1 but they got 4.2 instead but they are not aware of that. If the error message they see when they do `install(version="3.13")` was as clear as in the other situations, maybe they wouldn't get as confused as they seem to be right now.

In other words, I don't think that they _purposely_ installed a version of R that is too new for BioC 3.13, so we need to be careful to not make it too easy for people to use the wrong version of R for their version of Bioconductor. With your current proposal there's the risk that many people (typically developers) will install R 4.2 to work on BioC 3.13 or BioC 3.14 (note that BioC 3.14 will also be tight to R 4.1). This will cause problems. If people want to experiment with future R maybe they should invoke something else, say `install(version="future")`, instead of `install(version="3.13")`?

**Comment by LiNk-NY** [comment]
Thanks Hervé @hpages, 
I agree that we should make it harder for the average user to have the wrong setup. 
What I hope is that anyone building R-devel knows what they're doing enough to revert back to 4.1 if need be. 

We have to keep in mind that `BiocManager` still has to work with the environments it is built in.
As a CRAN package, it cannot be failing on R 4.2 (devel). 

That is an interesting option `version = "future"` and I will give this more thought. 

Best,
Marcel

**Comment by hpages** [comment]
> We have to keep in mind that BiocManager still has to work with the environments it is built in.

Maybe this is just a matter of skipping examples and tests that fail on "future R"?

**Comment by HenrikBengtsson** [comment]
I think a design goal should be that:

```r
BiocManager::install("BiocGenerics")
```

will always work regardless of what R version you're running.  Ideally, this should also be the case even if you stop updating/maintaining **BiocManager**.  In other words, it should be up to **BiocManager** to make sure this works, not the user.

**Comment by LiNk-NY** [comment]
Time to switch to R version 4.1 RC
